File: Sources/ThesisSkipApp/utils/Result.swift
//
//  APIResource.swift
//  MADiOS
//
//  Created by Bram Rampelberg on 03/01/2025.
//  Copyright © 2025 HOGENT. All rights reserved.
//

import Foundation

enum Result<T: Sendable>: Sendable {
    case success(data: T)
    case failure(cause: String, error: Error? = nil)

    #if SKIP
        private init(failureAndroid cause: String, error: Error? = nil) {
            AppLogger.error(
                "Failure created with cause: \(cause), error: \(error?.localizedDescription ?? "None")"
            )
            self = .failure(cause: cause, error: error)
        }

        static func failureWithLog(cause: String, error: Error? = nil) -> Result
        {
            return Result(failure: cause, error: error)
        }
    #else
        private init(
            failure cause: String,
            error: Error? = nil,
            file: String = #file,
            function: String = #function,
            line: Int = #line
        ) {
            AppLogger.error(
                "Failure created with cause: \(cause), error: \(error?.localizedDescription ?? "None")",
                file: file,
                function: function,
                line: line
            )
            self = .failure(cause: cause, error: error)
        }

        static func failureWithLog(
            cause: String,
            error: Error? = nil,
            file: String = #file,
            function: String = #function,
            line: Int = #line
        ) -> Result {
            return Result(
                failure: cause,
                error: error,
                file: file,
                function: function,
                line: line
            )
        }
    #endif

    var isSuccess: Bool {
        switch self {
        case .success: return true
        case .failure: return false
        }
    }

    var isFailure: Bool {
        !isSuccess
    }

    var data: T? {
        switch self {
        case .success(let data): return data
        default: return nil
        }
    }

    var failureCause: String? {
        switch self {
        case .failure(let cause, _): return cause
        default: return nil
        }
    }

    var error: Error? {
        switch self {
        case .failure(_, let error): return error
        default: return nil
        }
    }
}

File: .build/plugins/outputs/thesis-skip-app/ThesisSkipApp/destination/skipstone/ThesisSkipApp/src/main/kotlin/thesis/skip/app/Result.kt
//
//  APIResource.swift
//  MADiOS
//
//  Created by Bram Rampelberg on 03/01/2025.
//  Copyright © 2025 HOGENT. All rights reserved.
//

package thesis.skip.app

import skip.lib.*

import skip.foundation.*

internal sealed class Result<out T> {
    class SuccessCase<T>(val associated0: T): Result<T>() {
        val resultData = associated0
    }
    class FailureCase(val associated0: String, val associated1: Error?): Result<Nothing>() {
        val cause = associated0
        val failureError = associated1
    }


    internal val isSuccess: Boolean
        get() {
            when (this) {
                is Result.SuccessCase -> return true
                is Result.FailureCase -> return false
            }
        }

    internal val isFailure: Boolean
        get() = !isSuccess

    internal val data: T?
        get() {
            when (this) {
                is Result.SuccessCase -> {
                    val data = this.associated0
                    return data
                }
                else -> return null
            }
        }

    internal val failureCause: String?
        get() {
            when (this) {
                is Result.FailureCase -> {
                    val cause = this.associated0
                    return cause
                }
                else -> return null
            }
        }

    internal val error: Error?
        get() {
            when (this) {
                is Result.FailureCase -> {
                    val error = this.associated1
                    return error
                }
                else -> return null
            }
        }

    companion object {
        fun <T> success(resultData: T): Result<T> = SuccessCase(resultData)
        fun failure(cause: String, failureError: Error? = null): Result<Nothing> = FailureCase(cause, failureError)

        fun <T> init(failureAndroid: String, error: Error? = null): Result<T> {
            val cause = failureAndroid
            AppLogger.error("Failure created with cause: ${cause}, error: ${error?.localizedDescription ?: "None"}")
            return Result.failure(cause = cause, failureError = error)
        }
    }
}
internal fun <T> Result(failureAndroid: String, error: Error? = null): Result<T> {
    val cause = failureAndroid
    return Result.init(failureAndroid = cause, error = error)
}


internal fun <T> failureWithLog(cause: String, error: Error? = null): Result<T> {
    AppLogger.error("Failure created with cause: ${cause}, error: ${error?.localizedDescription ?: "None"}")
    return Result.failure(cause = cause, failureError = error)
}

